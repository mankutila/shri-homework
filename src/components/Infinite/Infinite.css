.loader {
  font-size: 10px;
  margin: 50px auto;
  text-indent: -9999em;
  width: 50px;
  height: 50px;
  border-radius: 50%;
  background: #57906a;
  background: linear-gradient(to right, #57906a 10%, rgba(255, 255, 255, 0) 42%);
  position: relative;
  animation: load3 1.4s infinite linear;
  transform: translateZ(0);
}
.loader:before {
  width: 50%;
  height: 50%;
  background: #57906a;
  border-radius: 100% 0 0 0;
  position: absolute;
  top: 0;
  left: 0;
  content: '';
}
.loader:after {
  background: #f5f9e9;
  width: 75%;
  height: 75%;
  border-radius: 50%;
  content: '';
  margin: auto;
  position: absolute;
  top: 0;
  left: 0;
  bottom: 0;
  right: 0;
}
@keyframes load3 {
  0% {
    -webkit-transform: rotate(0deg);
    transform: rotate(0deg);
  }
  100% {
    -webkit-transform: rotate(360deg);
    transform: rotate(360deg);
  }
}
