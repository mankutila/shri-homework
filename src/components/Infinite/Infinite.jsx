import React, {Component} from 'react';
import {connect} from 'react-redux';

import './Infinite.css';

const THRESHOLD = 500;

export class InfiniteComponent extends Component {

  constructor(props) {
    super(props);

    this.onScroll = this.onScroll.bind(this);
  }

  componentDidMount() {
    document.addEventListener('scroll', this.onScroll, {passive: true});
  }

  componentWillUnmount() {
    document.removeEventListener('scroll', this.onScroll);
  }

  onScroll() {
    if (!this.container || this.props.loading) {
      return;
    }

    let scrollTop = document.body.scrollTop || document.documentElement.scrollTop,
      containerHeight = this.container.clientHeight,
      windowHeight = window.innerHeight;

    if (scrollTop + windowHeight >= containerHeight - THRESHOLD && !this.props.allLoaded) {
      this.nextPage();
    }
  }

  async nextPage() {
    try {
      await this.props.fetchImages();
    } catch(error) {
      this.props.dispatch({
        type: 'LOAD_ERROR',
        error
      });
    }

  }

  render() {
    return (
      <div ref={(container) => this.container = container}>

        {this.props.children}

        {this.props.loading && (
          <div className="loading">
            <div className="loader">Загрузка...</div>
          </div>
        )}

      </div>
    );
  }
}

const stateToProps = (state) => ({
  loading: state.loading,
  allLoaded: state.allLoaded,
  error: state.error
});

export const Infinite = connect(stateToProps)(InfiniteComponent);