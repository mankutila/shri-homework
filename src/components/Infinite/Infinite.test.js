import React from 'react';
import { shallow } from 'enzyme';
import { createMockStore } from 'redux-test-utils';
import { Infinite } from './Infinite';

const shallowWithStore = (component, store) => {
  const context = {
    store,
  };
  return shallow(component, { context });
};

it('renders correctly', () => {

  const testState = {
    images: {
      images: [],
      page: 1,
      allLoaded: false
    },
    lightbox: {
      isOpen: false
    },
    tags: {
      current: 'architecture'
    }

  };
  const store = createMockStore(testState);

  const wrapper = shallowWithStore(<Infinite />, store)

  expect(wrapper).toMatchSnapshot();
});
