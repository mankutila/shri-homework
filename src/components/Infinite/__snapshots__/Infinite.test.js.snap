// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<InfiniteComponent
  allLoaded={false}
  current="architecture"
  dispatch={[Function]}
/>
`;
